src/ethereum/genesis.py

```diff
@@ -0,0 +1,84 @@
+"""
+Genesis Config Related Functionalities
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+..contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Functionalities to load the genesis configs and the configurations for
+different chains.
+"""
+import json
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Dict
+
+from ethereum.base_types import U256, Bytes, Bytes8, Uint, slotted_freezable
+from ethereum.frontier.eth_types import Address
+
+from .utils import (
+    hex_to_address,
+    hex_to_bytes,
+    hex_to_bytes8,
+    hex_to_u256,
+    hex_to_uint,
+)
+
+
+@slotted_freezable
+@dataclass
+class GenesisConfig:
+    """
+    Genesis Config has the alloc data for the pre-sale of ether and some
+    fields of the genesis block.
+    """
+
+    difficulty: Uint
+    extra_data: Bytes
+    gas_limit: Uint
+    nonce: Bytes8
+    timestamp: U256
+    alloc: Dict[Address, Dict[str, U256]]
+
+
+def load_genesis_config(genesis_file: str) -> GenesisConfig:
+    """
+    Obtain the genesis config, which includes the alloc data, based on the
+    given genesis json file.
+
+    Parameters
+    ----------
+    genesis_file :
+        The json file which contains the parameters for the genesis block
+        and the alloc data.
+
+    Returns
+    -------
+    genesis_config : `GenesisConfig`
+        The genesis config obtained from the json genesis file.
+    """
+    with open(genesis_file, "r") as genesis_file_handler:
+        genesis_data = json.load(genesis_file_handler)
+
+    alloc_data = {
+        hex_to_address(address): {"balance": hex_to_u256(account["balance"])}
+        for address, account in genesis_data["alloc"].items()
+    }
+
+    return GenesisConfig(
+        difficulty=hex_to_uint(genesis_data["difficulty"]),
+        extra_data=hex_to_bytes(genesis_data["extraData"]),
+        gas_limit=hex_to_uint(genesis_data["gasLimit"]),
+        nonce=hex_to_bytes8(genesis_data["nonce"]),
+        timestamp=hex_to_u256(genesis_data["timestamp"]),
+        alloc=alloc_data,
+    )
+
+
+ASSETS_DIR = Path(__file__).parent / "assets"
+
+MAINNET_GENESIS_CONFIG = load_genesis_config(f"{ASSETS_DIR}/mainnet.json")
```

Comment:
Constants are evaluated at import time, and exist forever. In the interest of speeding up imports, I think we should just call `genesis_configuration()` directly in the code that needs it, instead of loading it during import.

Reply:
Going forward while we support multiple chains and custom genesis blocks, I think we would need a configuration manager and we can directly call the `genesis_configuration` code over there. Is it ok if I mention this as a `todo` comment for the time being? Thanks.

Reply:
As long as we avoid loading the JSON at import time I'm pretty happy to merge.

Reply:
I have removed this from import time, and please consider this discussion resolved.

Though I completely agree that the import time would be more, I would like a chance to explain why I preferred loading the JSON at import time itself. Please find the minimal reproducer below.

### constants.py
```
from time import sleep

def load(val):
    sleep(10)
    return val

c1 = load(5)
c2 = load(7)
```

### a.py
```
from .constants import c1, c2

def hello():
    print("Calling function hello")
    print(c1)
    print(c2)
```

### b.py
```
from .constants import c1, c2

def hello1():
    print("Called function hello1")
    print(c1)
    print(c2)
```

### ipython output
```
In [1]: time from test_proj.a import hello                                                                                                                                                                         
CPU times: user 660 µs, sys: 3.74 ms, total: 4.4 ms
Wall time: 20 s

In [2]: time hello()                                                                                                                                                                                               
Calling function hello
5
7
CPU times: user 70 µs, sys: 8 µs, total: 78 µs
Wall time: 84.9 µs

In [3]: time from test_proj.b import hello1                                                                                                                                                                        
CPU times: user 1.24 ms, sys: 132 µs, total: 1.37 ms
Wall time: 1.09 ms

In [4]: time hello1()                                                                                                                                                                                              
Called function hello1
5
7
CPU times: user 278 µs, sys: 40 µs, total: 318 µs
Wall time: 232 µs
```

## Doubt
Hence as per above reproducer, it doesn't seem like there is a difference between loading the genesis wherever needed and loading the genesis as a constant and importing it. Rather making the genesis as a constant and importing it wherever needed is actually memorizing it automatically. That's why I initially loaded the config as a constant. Given that at this point of time, there are no multiple usages of this config, I'm removing it. Please let me know if I am incorrect anywhere, I'm just posting this to improve my knowledge. Thanks.

Reply:
On line `[1]` of your output, we see that _just importing_ `hello` took 20 seconds of wall time.

That's exactly what we want to avoid. Someone who's simply importing the software and not calling any functions shouldn't have to pay the cost of loading the genesis data.

Reply:
Got it, thanks Sam.